Add an option to use a popup menu for code actions

dotdash opened the pull request:
For code actions, some people might find it more convenient to use a popup menu that shows up at the cursor position, instead of using the quickpick window at the bottom of the screen.

![code_action_popup](https://user-images.githubusercontent.com/230962/192136305-0cb75b73-a32a-43ec-93bd-30b5129ab9e4.jpg)

prabirshrestha commented:
Thanks for the PR. Could you use Vital's Popup api instead as it will allow us to easily maintain code for both vim and neovim. You can refer to the example [here](https://github.com/prabirshrestha/vim-lsp/blob/master/autoload/lsp/internal/document_hover/under_cursor.vim). If you don't have neovim don't worry I can test it out.

Instead of having a boolean can you use `--ui=` flag that we have been moving to. 
https://github.com/prabirshrestha/vim-lsp/blob/4d8a27513daef41915eaf2d315a2662e95d71080/doc/vim-lsp.txt#L1656-L1672. This will also allow us to have multiple mapping. https://github.com/prabirshrestha/vim-lsp/blob/4d8a27513daef41915eaf2d315a2662e95d71080/doc/vim-lsp.txt#L1884-L1886

dotdash commented:
Are you referring to https://github.com/prabirshrestha/vim-lsp/blob/master/autoload/vital/_lsp/VS/Vim/Window/FloatingWindow.vim ? If so, I don't see any support for `popup_menu` or the corresponding options like `filter` and `cursorline` to manually configure the popup window. Am I looking at the wrong place, or do you want me to extend the existing API?

prabirshrestha commented:
For now should be good to directly use it. But it would be good if you can support `--ui` flag instead.

prabirshrestha commented:
Is it possible to create a util method in [internal/ui](https://github.com/prabirshrestha/vim-lsp/tree/master/autoload/lsp/internal/ui). This will allow us to easily support neovim in the future.

```vim
call lsp#lsp#internal#ui#popupmenu#open({
  \   'title': '...',
  \   'items': [],
  \   'callback': function('...'),
  \   'line': 1,
  \   'col': 1,  
  \ })
```

Also would be good to support number as prefix so that one can easily select 1 or 2. We probably want to only support numbers from `1-9` and beyond that it won't include number prefix.

```
[1] [rust-analyzer] Import serde::Deserialize
[2] [rust-analyzer] Qualify as serde::Deserialize
```

Also noticed that markdown content is not applied. You can refer to example at https://github.com/prabirshrestha/vim-lsp/blob/master/autoload/lsp/internal/completion/documentation.vim#L4.

<img width="747" alt="image" src="https://user-images.githubusercontent.com/287744/192196803-b40a6c49-4d68-4041-8c7a-6f1144ac194d.png">

I like this for code action. We can probably make this default at some point in the master branch once it has baked for few weeks.

dotdash commented:
If switched to using the `--ui` flag, which needed an extension to the arg parser. I didn't get around to do the markdown, numbering and wrapper function yet. I hope to get that done next week.

prabirshrestha commented on autoload/lsp/ui/vim/code_action.vim:
```diff
@@ -1,5 +1,7 @@
 " vint: -ProhibitUnusedVariable
 
+let s:popup_menu_is_available = exists('*popup_menu')
```
we should move this check here. https://github.com/prabirshrestha/vim-lsp/blob/master/autoload/lsp/utils.vim. Old distros that has incomplete patches can be disabled in future if we see others reporting issues.

prabirshrestha reviewed (changes requested):
There are few more lint issues.

I prefer to use `ctrl+j/k` for navigation instead of `ctrl+n/p`. Would be good to support both. Can we add support for `lsp_float_opened` and `lsp_float_closed`. We can also autoregister if plugs are not enabled. https://github.com/prabirshrestha/vim-lsp/blob/74e458bc9c7532ff1959b2443a1d73494e3673b6/autoload/lsp/internal/document_hover/under_cursor.vim#L220-L226

dotdash commented:
> Also would be good to support number as prefix so that one can easily select 1 or 2.

Added a filter function to handle that. I've integrated it into the `lsp#internal#ui#popupmenu#open` function as I expected you to want that functionality on future popup menu additions as well.

> I prefer to use ctrl+j/k for navigation instead of ctrl+n/p. Would be good to support both.

The default keybindings provided by vim already include plain j/k keys for moving down/up, I've added C-j/C-k by means to the filter function as well now.

> Can we add support for lsp_float_opened and lsp_float_closed

Added.

> We can also autoregister if plugs are not enabled. 

I'm sorry, I'm not sure I understand what you mean by that. Do you mean the `<Plug>(lsp-float-close)` mapping? If so, mappings are disabled on popup menus by default. And as keys are processed individually by the filter function, I'm not sure how to properly make use of the mapping anyway. Suggestions welcome.

prabirshrestha commented:
In that case we probably want to manually add the bindings and explicitly mention in docs that `lsp-float-opened/closed` events are not supported.

dotdash commented:
I have added the `doautocommand` calls for the `lsp-float-opened/closed` events. `lsp-float-opened` is triggered after the menu has been opened and `lsp-float-closed` is triggered in the callback, which is right before the window actually gets closed, but that's the best I could do. Did I misunderstand how they are meant to be used?

What do you mean by "manually adding the bindings"? For clarification, the default `popup_filter_menu()` function has these keybindings:
```		    j <Down> <C-N>	select item below
		    k <Up> <C-P>	select item above
		    <Space> <Enter>	accept current selection
		    x Esc CTRL-C	cancel the menu
```

The custom filter I created adds `<C-j>` and `<C-k>` as per your request, as well as the `1`-`9` shortcuts. 
What other bindings do you have in mind?

prabirshrestha commented:
Thanks for the PR. All looks good.
